### ISSUE TITLE
Enforcement of file checksum requirement

### ISSUE THREAD

**Issue by karlcz** [nih-cfde/cfde-deriva#250]
The C2M2 spec says that at least one file checksum (md5 or sha256) is required, but the current frictionless spec and translated relational schema have not enforced this constraint affecting multiple columns.  Some file records in current data violate this spec requirement.

We want to add a custom validator to make ingest of future submissions more strict, rejecting such submissions with a meaningful error message.

However, we need to allow the next release to be built with submissions already received, which exhibit this problem. We may need a feature flag or other control to allow associated release processing...?

**Comment by abradyIGS** [comment]
If I close one eye I agree with that, but I've confused myself with some possible ambiguity in phrases like "January update," so let me write two timelines and ask you to pick one. Timeline 1:
* you announce now that submissions received during Q3 must be checksum-compliant
* i will need to get that validator ready by the first few weeks in November at the latest (not end of January) so they can use it
* the January update (comprised of incoming data received during Q3) would thus be compliant
Timeline 2:
* you announce now that submissions received during Q4 must be checksum-compliant
* validator would be ready by end of January as you say
* the _May_ update (comprised of incoming data received during Q4) would thus be compliant
I think I want Timeline 1, but I don't know all the other process constraints (or if that constitutes a full cycle's worth of warning).

**Comment by ACharbonneau** [comment]
Sorry. Yes I meant timeline 1. I was talking entirely from the submitter perspective. So they would have the rest of this cycle, plus the next full cycle, to make a new thing, to submit in January. As opposed to what we typically do which is give them new directives at the beginning of a cycle...so I'd like to tell them now, not in early November.

**Comment by ACharbonneau** [comment]
I don't know whether Karl would have the validator working in the portal by Nov, but I feel like we could announce the requirement, say its going into effect for January, manually validate that one time, and make the software actually compliant at the January release, if needed. 

Hopefully that sentence is sort of sensical

**Comment by abradyIGS** [comment]
For a client-side compliance check, a one-off script is easy, I can have that ready for something like a beta distro at the beginning of November so DCCs can make sure they're fine before submitting. Can refine and maybe merge with the term-builder as "The Client-Side Thing" during Q3.

As for HuBMAP, we'll need to meet separately and probably more than once to disentangle all the current concerns there in a way that doesn't result in too many exploding heads.

**Comment by karlcz** [comment]
I've got a version of this in the ingest process now, so as soon as we deploy it the action provider becomes more strict about hashes. The final rule as per 2021-10-07 C2M2 modeling WG, is that a checksum is required if a file's persistent\_id is popuated.

I also added a feature switch via environment variable, to allow this code to be used by an admin to manually rebuild a submission and/or release while bypassing the extra enforcement check. (I.e. if we want to allow a "legacy" dataset to pass further through the system as an interim measure, without having to roll back any code.)
